src/python/pants/backend/python/dependency_inference/import_parser.py

```diff
@@ -35,40 +34,55 @@ def all_imports(self) -> FrozenOrderedSet[str]:
         return FrozenOrderedSet(sorted([*self.explicit_imports, *self.inferred_imports]))
 
 
-def parse_file(*, filename: str, content: str) -> Optional[Tuple]:
+_Visitor = TypeVar("_Visitor")
+
+
+class VisitorInterface:
+    explicit_imports: Set[str]
+    inferred_imports: Set[str]
+
+    def __init__(self, module_name: str) -> None:
+        ...
+
+    @classmethod
+    def visit_tree(cls: Type[_Visitor], tree: Any, module_name: str) -> _Visitor:
+        ...
+
+
+def parse_file(*, filename: str, content: str, module_name: str) -> VisitorInterface:
+    """Parse the file for python imports, and return a visitor with the imports it found."""
     try:
-        # NB: The Python 3 ast is generally backwards-compatible with earlier versions. The only
-        # breaking change is `async` `await` becoming reserved keywords in Python 3.7 (deprecated
-        # in 3.6). If the std-lib fails to parse, we could use typed-ast to try parsing with a
-        # target version of Python 3.5, but we don't because Python 3.5 is almost EOL and has very
-        # low usage.
-        # We will also fail to parse Python 3.8 syntax if Pants is run with Python 3.6 or 3.7.
-        # There is no known workaround for this, beyond users changing their `./pants` script to
-        # always use >= 3.8.
-        tree = ast3.parse(content, filename=filename)
-        visitor_cls = _Py3AstVisitor if sys.version_info[:2] < (3, 8) else _Py38AstVisitor
-        return tree, visitor_cls
+        py27_tree = ast27.parse(content, filename=filename)
```

Comment:
It was intentional to parse with Py3 before attempting Py2. The reasoning is that we expect the vast majority of user code to be Py3, so we can avoid unnecessary work by parsing Py3 first. (We'd rather optimize for the Py3 experience than Py2 experience)

I realize that should have been left in a comment, sorry about that. I'd appreciate if you could please flip back the order and perhaps add a brief comment.

Reply:
Left a comment with a TODO pointing to #10921!